packages/langium/src/parser/lexer.ts

```diff
@@ -66,6 +67,45 @@ export class DefaultLexer implements Lexer {
     }
 }
 
+/**
+ * A lexer that is aware of indentation in the input text.
+ * The only purpose of this lexer is to reset the internal state of the {@link IndentationAwareTokenBuilder}
+ * between the tokenization of different text inputs.
+ *
+ * In your module, you can override the default lexer with this one as such:
+ * ```ts
+ * parser: {
+ *    TokenBuilder: () => new IndentationAwareTokenBuilder(),
+ *    Lexer: (services) => new IndentationAwareLexer(services),
+ * }
+ * ```
+ */
+export class IndentationAwareLexer extends DefaultLexer {
+    private indentationTokenBuilder?: IndentationAwareTokenBuilder;
+
+    constructor(services: LangiumCoreServices) {
+        super(services);
+        if (services.parser.TokenBuilder instanceof IndentationAwareTokenBuilder) {
+            this.indentationTokenBuilder = services.parser.TokenBuilder;
```

Comment:
Suggestion: I believe we should make it very clear that this lexer is supposed to be used in conjunction with the `IndentationAwareTokenBuilder` (or an extension of it). Given that every other combination is pretty nonsensical, we should throw an error if we find that the token builder is not of type `IndentationAwareTokenBuilder`.

This also allows us to make the `indentationTokenBuilder` field non-optional.